File: Easy/4. Tournament Winner/tournamentWinner.go
package main

import "fmt"

const HOME_TEAM_WON = 1

// TournamentWinner function
func TournamentWinner(competitions [][]string, results []int) string {
	currentBestTeam := ""
	scores := map[string]int{currentBestTeam: 0}

	for i, competition := range competitions {
		result := results[i]
		homeTeam, awayTeam := competition[0], competition[1]

		winningTeam := awayTeam
		if result == HOME_TEAM_WON {
			winningTeam = homeTeam
		}

		updateScores(winningTeam, 3, scores)

		if scores[winningTeam] > scores[currentBestTeam] {
			currentBestTeam = winningTeam
		}
	}
	return currentBestTeam
}

// updateScores function
func updateScores(team string, points int, scores map[string]int) {
	scores[team] += points
}

func main() {
	competitions := [][]string{{"HTML", "C#"}, {"C#", "Python"}, {"Python", "HTML"}}
	results := []int{0, 0, 1}
	fmt.Println("Winner: ", TournamentWinner(competitions, results))
}

File: Easy/4. Tournament Winner/tournamentWinner.py
HOME_TEAM_WON = 1

# O(n) time | O(k) space
def tournamentWinner(competitions, results):
    currentBestTeam = ""
    scores = {currentBestTeam: 0}

    for i, competition in enumerate(competitions):
        result = results[i]
        homeTeam, awayTeam = competition
        
        winningTeam = homeTeam if result == HOME_TEAM_WON else awayTeam
        
        updateScores(winningTeam, 3, scores)
        
        if scores[winningTeam] > scores[currentBestTeam]:
            currentBestTeam = winningTeam
            
    return currentBestTeam

def updateScores(team, points, scores):
    if team not in scores:
        scores[team] = 0
        
    scores[team] += points
	
if __name__ == "__main__":
    competitions = [
    ["HTML", "C#"],
    ["C#", "Python"],
    ["Python", "HTML"]
    ]
    results = [0, 0, 1]
	
    print(f"Winner: {tournamentWinner(competitions, results)}")
